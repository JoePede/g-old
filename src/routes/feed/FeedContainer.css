.container {
  padding: 40px 0 0;

  > div:not(:last-child) {
    margin-bottom: 2rem;
  }
}
