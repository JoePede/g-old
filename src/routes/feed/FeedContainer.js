import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { loadFeed } from '../../actions/feed';
import {
  getActivities,
  getFeedIsFetching,
  getFeedErrorMessage,
} from '../../reducers/index';
import FetchError from '../../components/FetchError';
import Activity from '../../components/Activity';

class FeedContainer extends React.Component {
  static propTypes = {
    activities: PropTypes.arrayOf(
      PropTypes.shape({
        id: PropTypes.string.isRequired,
      })
    ).isRequired,
    isFetching: PropTypes.bool.isRequired,
    loadFeed: PropTypes.func.isRequired,
    errorMessage: PropTypes.string,
  };

  static defaultProps = {
    errorMessage: '',
  };
  isReady() {
    // Probably superflue bc we are awaiting the LOAD_PROPOSAL_xxx flow
    return this.props.activities != null;
  }

  render() {
    const { activities, isFetching, errorMessage } = this.props;
    if (isFetching && !activities.length) {
      return <p> Loading ... </p>;
    }

    if (errorMessage && !activities.length) {
      return (
        <FetchError
          message={errorMessage}
          onRetry={() => this.props.loadFeed()}
        />
      );
    }

    const outDated = {};
    const polls = {};
    const proposals = {};
    return (
      <div>
        <h1>feed</h1>
        {activities.map(activity => {
          if (!activity.object) return null; // hotfix - problem is ondelete cascade for votes
          if (activity.type === 'statement') {
            if (activity.verb === 'update') {
              if (activity.objectId in outDated) {
                return null;
              }
              if (
                activity.object.pollId in polls &&
                activity.object.author.id === polls[activity.object.pollId]
              ) {
                return null;
              }
              outDated[activity.objectId] = activity.objectId;
              polls[activity.object.pollId] = activity.object.author.id;
            } else if (activity.verb === 'delete') {
              outDated[activity.objectId] = activity.object.author.id;
              polls[activity.object.pollId] = activity.object.author.id;

              return null;
            } else if (activity.objectId in outDated) {
              return null;
            } else if (activity.verb === 'create') {
              if (activity.object) {
                if (
                  activity.object.pollId in outDated &&
                  activity.object.author.id === outDated[activity.object.pollId]
                ) {
                  return null;
                }
                if (
                  activity.object.pollId in polls &&
                  activity.object.author.id === polls[activity.object.pollId]
                ) {
                  return null;
                }
                outDated[activity.object.pollId] = activity.object.author.id;
              } else {
                return null;
              }
            }
          } else if (activity.type === 'proposal') {
            if (activity.objectId in proposals) {
              return null;
            }
            proposals[activity.objectId] = activity.objectId;
          }

          return (
            <Activity
              key={activity.id}
              actor={activity.actor}
              date={activity.createdAt}
              verb={activity.verb}
              content={activity.object}
            />
          );
        })}
      </div>
    );
  }
}
// TODO implement memoiziation with reselect
const mapStateToProps = state => ({
  activities: getActivities(state, 'all'),
  isFetching: getFeedIsFetching(state, 'all'),
  errorMessage: getFeedErrorMessage(state, 'all'),
});

const mapDispatch = {
  loadFeed,
};

export default connect(mapStateToProps, mapDispatch)(FeedContainer);
