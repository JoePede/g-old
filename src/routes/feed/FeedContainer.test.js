/**
 * React Starter Kit (https://www.reactstarterkit.com/)
 *
 * Copyright Â© 2014-present Kriasoft, LLC. All rights reserved.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE.txt file in the root directory of this source tree.
 */

/* eslint-env jest */
/* eslint-disable padded-blocks, no-unused-expressions */

import React from 'react';
import renderer from 'react-test-renderer';
import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import { IntlProvider } from 'react-intl';

import FeedContainer from './FeedContainer';
import App from '../../components/App';

const middlewares = [thunk.withExtraArgument({ fetch: () => {} })];
const mockStore = configureStore(middlewares);
const initialState = {
  runtime: {
    availableLocales: ['de-DE'],
  },
  intl: {
    locale: 'de-DE',
  },
  entities: {
    users: {
      byId: {
        1: {
          id: '1',
          name: 'Grace',
          surname: 'Von',
          avatar: 'https://api.adorable.io/avatars/32/GraceVon.io.png',
        },
      },
    },
    votes: {
      1: {
        id: '1',
        position: 'con',
      },
    },
    statements: {
      byId: {
        1: {
          __typename: 'StatementDL',
          id: '1',
          likes: 0,
          text: 'before',
          pollId: '1',
          createdAt: 'Sat Jul 29 2017 23:52:50 GMT+0200 (CEST)',
          updatedAt: null,
          vote: '1',
          author: '1',
        },
      },
    },

    proposals: {
      byId: {
        1: { __typename: 'ProposalDL' },
      },
    },
    activities: {
      allIds: { ids: [1] },
      byId: {
        1: {
          id: '1',
          type: 'statement',
          objectId: '1',
          verb: 'create',
          createdAt: 'Sat Jun 17 2017 09:09:50 GMT+0200 (CEST)',
          actor: '1',
          object: {
            id: '1',
            schema: 'StatementDL',
          },
        },
      },
    },
  },
};

const updatedEntities = {
  entities: {
    users: {
      byId: {
        1: {
          id: '1',
          name: 'Grace',
          surname: 'Von',
          avatar: 'https://api.adorable.io/avatars/32/GraceVon.io.png',
        },
      },
    },
    votes: {
      1: {
        id: '1',
        position: 'con',
      },
    },
    statements: {
      byId: {
        1: {
          __typename: 'StatementDL',
          id: '1',
          likes: 0,
          text: 'after',
          pollId: '1',
          createdAt: 'Sat Jul 29 2017 23:52:50 GMT+0200 (CEST)',
          updatedAt: 'Sat Jul 29 2017 23:52:51 GMT+0200 (CEST)',
          vote: '1',
          author: '1',
        },
      },
    },

    proposals: {
      byId: {
        1: { __typename: 'ProposalDL' },
      },
    },
    activities: {
      allIds: { ids: [2, 1] },
      byId: {
        1: {
          id: '1',
          type: 'statement',
          objectId: '1',
          verb: 'create',
          createdAt: 'Sat Jun 17 2017 09:09:50 GMT+0200 (CEST)',
          actor: '1',
          object: {
            id: '1',
            schema: 'StatementDL',
          },
        },
        2: {
          id: '2',
          type: 'statement',
          objectId: '1',
          verb: 'update',
          createdAt: 'Sat Jun 17 2017 09:09:51 GMT+0200 (CEST)',
          actor: '1',
          object: {
            id: '1',
            schema: 'StatementDL',
          },
        },
      },
    },
  },
};
describe('FeedContainer', () => {
  it('Displays activity', () => {
    const store = mockStore(initialState);

    const wrapper = renderer
      .create(
        <IntlProvider>
          <App context={{ insertCss: () => {}, store }}>
            <FeedContainer />
          </App>
        </IntlProvider>,
      )
      .toJSON();

    expect(wrapper).toMatchSnapshot();
  });

  it('Hides outdated content', () => {
    const updatedStore = {
      ...initialState,
      entities: {
        ...updatedEntities.entities,
      },
    };
    const store = mockStore(updatedStore);

    const wrapper = renderer
      .create(
        <IntlProvider>
          <App context={{ insertCss: () => {}, store }}>
            <FeedContainer />
          </App>
        </IntlProvider>,
      )
      .toJSON();
    const asString = JSON.stringify(wrapper);
    expect(asString).toMatch(/after/);
  });

  it('Dont displays deleted statements', () => {
    const updatedStore = {
      ...initialState,
      entities: {
        ...updatedEntities.entities,
        activities: {
          allIds: { ids: [3, 2, 1] },
          byId: {
            ...updatedEntities.entities.activities.byId,
            3: {
              id: '3',
              type: 'statement',
              objectId: '1',
              verb: 'delete',
              createdAt: 'Sat Jun 17 2017 09:09:52 GMT+0200 (CEST)',
              actor: '1',
              object: {
                id: '1',
                schema: 'StatementDL',
              },
            },
          },
        },
      },
    };
    const store = mockStore(updatedStore);

    const wrapper = renderer
      .create(
        <IntlProvider>
          <App context={{ insertCss: () => {}, store }}>
            <FeedContainer />
          </App>
        </IntlProvider>,
      )
      .toJSON();
    const asString = JSON.stringify(wrapper);
    expect(asString).not.toMatch(/after/);
  });

  it('Merges statements from same user on same poll', () => {
    const updatedStore = {
      ...initialState,
      entities: {
        ...updatedEntities.entities,
        statements: {
          byId: {
            ...updatedEntities.entities.statements.byId,
            2: {
              __typename: 'StatementDL',
              id: '2',
              likes: 0,
              text: 'next',
              pollId: '1',
              createdAt: 'Sat Jul 30 2017 23:52:50 GMT+0200 (CEST)',
              updatedAt: 'Sat Jul 30 2017 23:52:51 GMT+0200 (CEST)',
              vote: '2',
              author: '1',
            },
          },
        },

        votes: {
          ...updatedEntities.entities.votes,
          2: {
            id: '2',
            position: 'pro',
          },
        },
        activities: {
          allIds: { ids: [3, 2, 1] },
          byId: {
            ...updatedEntities.entities.activities.byId,
            3: {
              id: '3',
              type: 'statement',
              objectId: '2',
              verb: 'create',
              createdAt: 'Sat Jun 17 2017 10:09:52 GMT+0200 (CEST)',
              actor: '1',
              object: {
                id: '2',
                schema: 'StatementDL',
              },
            },
          },
        },
      },
    };
    let store = mockStore(updatedStore);

    let wrapper = renderer
      .create(
        <IntlProvider>
          <App context={{ insertCss: () => {}, store }}>
            <FeedContainer />
          </App>
        </IntlProvider>,
      )
      .toJSON();
    let asString = JSON.stringify(wrapper);
    expect(asString).toMatch(/next/);

    const deleteLastStatementAfterVoteChangeStore = {
      ...updatedStore,
      entities: {
        ...updatedStore.entities,
        activities: {
          allIds: { ids: [4, 3, 2, 1] },
          byId: {
            ...updatedStore.entities.activities.byId,
            4: {
              id: '4',
              type: 'statement',
              objectId: '2',
              verb: 'delete',
              createdAt: 'Sat Jun 17 2017 10:09:52 GMT+0200 (CEST)',
              actor: '1',
              object: {
                id: '2',
                schema: 'StatementDL',
              },
            },
          },
        },
      },
    };

    store = mockStore(deleteLastStatementAfterVoteChangeStore);

    wrapper = renderer
      .create(
        <IntlProvider>
          <App context={{ insertCss: () => {}, store }}>
            <FeedContainer />
          </App>
        </IntlProvider>,
      )
      .toJSON();
    asString = JSON.stringify(wrapper);
    expect(asString).not.toMatch(/next/);
    expect(asString).not.toMatch(/after/);

    // Test update of retracted

    const updateLastStatementAfterVoteChangeStore = {
      ...updatedStore,
      entities: {
        ...updatedStore.entities,
        statements: {
          byId: {
            ...updatedStore.entities.statements.byId,
            3: {
              __typename: 'StatementDL',
              id: '3',
              likes: 0,
              text: 'newest',
              pollId: '1',
              createdAt: 'Sat Jul 30 2017 23:52:50 GMT+0200 (CEST)',
              updatedAt: 'Sat Jul 30 2017 23:52:51 GMT+0200 (CEST)',
              vote: '2',
              author: '1',
            },
          },
        },
        activities: {
          allIds: { ids: [4, 3, 2, 1] },
          byId: {
            ...updatedStore.entities.activities.byId,
            4: {
              id: '4',
              type: 'statement',
              objectId: '2',
              verb: 'update',
              createdAt: 'Sat Jun 17 2017 10:09:52 GMT+0200 (CEST)',
              actor: '1',
              object: {
                id: '3',
                schema: 'StatementDL',
              },
            },
          },
        },
      },
    };

    store = mockStore(updateLastStatementAfterVoteChangeStore);

    wrapper = renderer
      .create(
        <IntlProvider>
          <App context={{ insertCss: () => {}, store }}>
            <FeedContainer />
          </App>
        </IntlProvider>,
      )
      .toJSON();
    asString = JSON.stringify(wrapper);
    expect(asString).not.toMatch(/next/);
    expect(asString).toMatch(/newest/);
  });
});
