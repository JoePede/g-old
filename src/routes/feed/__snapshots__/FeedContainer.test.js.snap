// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FeedContainer Displays activity 1`] = `
<div>
  <h1>
    feed
  </h1>
  <div
    className="container"
  >
    <span>
      vor 2 Monaten
    </span>
    <h1>
      Look at that statement!
    </h1>
    <div>
      <a
        href="/proposal/xxx/1"
        onClick={[Function]}
      >
        <div
          className="rootAlt contra"
        >
          <img
            alt="IMG"
            className="avatar"
            src="https://api.adorable.io/avatars/32/GraceVon.io.png"
          />
          <div
            style={
              Object {
                "width": "100%",
              }
            }
          >
            <div
              className="header"
            >
              <div>
                <span
                  className="author"
                >
                  Grace
                   
                  Von
                </span>
                <span>
                  
                </span>
              </div>
            </div>
            <div
              className="text"
            >
              before
            </div>
          </div>
        </div>
      </a>
    </div>
  </div>
</div>
`;
